# Analysing Hacker News Posts

We shall analyze hacker news posts specifically the ones that aim at asking the Hacker news community"Ask HN" and the ones that post to the community"Show HN".We shall compare these two posts to determine the following:
- Do Ask HN or Show HN receive more comments on average?
- Do posts created at a certain time receive more comments on average?


## Introduction
first we shall import the data and remove the header.It should also be noted here that the data set has been reduced from about 300000 rows to 20000 rows to remove submissions that did not receive any comments and randomly sampling from the remaining submission.

In [1]:
#Import csv and read in the data
import csv

In [2]:
Data = open("hacker_news.csv")

In [3]:
new_data = csv.reader(Data)

In [4]:
hn = list(new_data)

In [5]:
#View first five rows
hn[0:4]

[['id', 'title', 'url', 'num_points', 'num_comments', 'author', 'created_at'],
 ['12224879',
  'Interactive Dynamic Video',
  'http://www.interactivedynamicvideo.com/',
  '386',
  '52',
  'ne0phyte',
  '8/4/2016 11:52'],
 ['10975351',
  'How to Use Open Source and Shut the Fuck Up at the Same Time',
  'http://hueniverse.com/2016/01/26/how-to-use-open-source-and-shut-the-fuck-up-at-the-same-time/',
  '39',
  '10',
  'josep2',
  '1/26/2016 19:30'],
 ['11964716',
  "Florida DJs May Face Felony for April Fools' Water Joke",
  'http://www.thewire.com/entertainment/2013/04/florida-djs-april-fools-water-joke/63798/',
  '2',
  '1',
  'vezycash',
  '6/23/2016 22:20']]

In [6]:
headers = hn[0]

In [7]:
headers

['id', 'title', 'url', 'num_points', 'num_comments', 'author', 'created_at']

## Removing headers from the list

In [8]:
hn = hn[1:]

In [9]:
hn[0:4]

[['12224879',
  'Interactive Dynamic Video',
  'http://www.interactivedynamicvideo.com/',
  '386',
  '52',
  'ne0phyte',
  '8/4/2016 11:52'],
 ['10975351',
  'How to Use Open Source and Shut the Fuck Up at the Same Time',
  'http://hueniverse.com/2016/01/26/how-to-use-open-source-and-shut-the-fuck-up-at-the-same-time/',
  '39',
  '10',
  'josep2',
  '1/26/2016 19:30'],
 ['11964716',
  "Florida DJs May Face Felony for April Fools' Water Joke",
  'http://www.thewire.com/entertainment/2013/04/florida-djs-april-fools-water-joke/63798/',
  '2',
  '1',
  'vezycash',
  '6/23/2016 22:20'],
 ['11919867',
  'Technology ventures: From Idea to Enterprise',
  'https://www.amazon.com/Technology-Ventures-Enterprise-Thomas-Byers/dp/0073523429',
  '3',
  '1',
  'hswarna',
  '6/17/2016 0:01']]

## Extracting Ask HN and Show Hn

In [10]:
ask_posts =[]
show_posts = []
other_posts = []

for i in hn:
    title = i[1]
    title = title.lower()
    if title.startswith("ask hn"):
        ask_posts.append(i)
    elif  title.startswith("show hn"):
        show_posts.append(i)
    else:
        other_posts.append(i)
        
print("ask-posts =",  len(ask_posts))
print("show_posts =", len(show_posts))
print("other_posts =", len(other_posts))
      

ask-posts = 1744
show_posts = 1162
other_posts = 17194


In [11]:
#Lets view first rows for ask_post
ask_posts[0:4]

[['12296411',
  'Ask HN: How to improve my personal website?',
  '',
  '2',
  '6',
  'ahmedbaracat',
  '8/16/2016 9:55'],
 ['10610020',
  'Ask HN: Am I the only one outraged by Twitter shutting down share counts?',
  '',
  '28',
  '29',
  'tkfx',
  '11/22/2015 13:43'],
 ['11610310',
  'Ask HN: Aby recent changes to CSS that broke mobile?',
  '',
  '1',
  '1',
  'polskibus',
  '5/2/2016 10:14'],
 ['12210105',
  'Ask HN: Looking for Employee #3 How do I do it?',
  '',
  '1',
  '3',
  'sph130',
  '8/2/2016 14:20']]

In [12]:
#Lets view first rows for show_post
show_posts[0:4]

[['10627194',
  'Show HN: Wio Link  ESP8266 Based Web of Things Hardware Development Platform',
  'https://iot.seeed.cc',
  '26',
  '22',
  'kfihihc',
  '11/25/2015 14:03'],
 ['10646440',
  'Show HN: Something pointless I made',
  'http://dn.ht/picklecat/',
  '747',
  '102',
  'dhotson',
  '11/29/2015 22:46'],
 ['11590768',
  'Show HN: Shanhu.io, a programming playground powered by e8vm',
  'https://shanhu.io',
  '1',
  '1',
  'h8liu',
  '4/28/2016 18:05'],
 ['12178806',
  'Show HN: Webscope  Easy way for web developers to communicate with Clients',
  'http://webscopeapp.com',
  '3',
  '3',
  'fastbrick',
  '7/28/2016 7:11']]

## Lets Calculate the average number of post for Ask HN and Post HN

In [13]:
total_ask_comments = 0
for i in ask_posts:
    num_comments = i[4]
    num_comments = int(num_comments)
    total_ask_comments += num_comments
    
    avg_ask_comments = total_ask_comments/len(ask_posts)
print(avg_ask_comments)            

14.038417431192661


In [14]:
total_show_comments = 0
for i in show_posts:
    num_comments = i[4]
    num_comments = int(num_comments)
    total_show_comments += num_comments
    
    avg_show_comments = total_show_comments/len(show_posts)
print(avg_show_comments)  

10.31669535283993


## Now let's find the amount of Ask Posts and Comments by Hour Created

Next, we'll determine if we can maximize the amount of comments an ask post receives by creating it at a certain time. First, we'll find the amount of ask posts created during each hour of day, along with the number of comments those posts received. Then, we'll calculate the average amount of comments ask posts created at each hour of the day receive

In [15]:
import datetime as dt
result_list = []
for i in ask_posts:
    created = i[6]
    num_comm = int(i[4])
    result_list.append([created,num_comm])

counts_by_hour = {}
comments_by_hour = {}

for r in result_list:
    comments = r[1]
    hour = r[0]
    hour = dt.datetime.strptime(hour,"%m/%d/%Y %H:%M")
    hour = dt.datetime.strftime(hour,"%H")
    if hour not in counts_by_hour:
        counts_by_hour[hour] = 1
        comments_by_hour[hour] = comments
    else:
        counts_by_hour[hour] += 1
        comments_by_hour[hour] += comments
print(comments_by_hour)
print(counts_by_hour)
    
    

{'10': 793, '17': 1146, '22': 479, '09': 251, '05': 464, '01': 683, '13': 1253, '16': 1814, '18': 1439, '00': 447, '07': 267, '02': 1381, '20': 1722, '19': 1188, '12': 687, '11': 641, '03': 421, '06': 397, '23': 543, '08': 492, '15': 4477, '04': 337, '14': 1416, '21': 1745}
{'10': 59, '17': 100, '22': 71, '09': 45, '05': 46, '01': 60, '13': 85, '16': 108, '18': 109, '00': 55, '07': 34, '02': 58, '20': 80, '19': 110, '12': 73, '11': 58, '03': 54, '06': 44, '23': 68, '08': 48, '15': 116, '04': 47, '14': 107, '21': 109}


## Calculate the average number of Ask HN by the hour

In [16]:
avg_by_hour = []
for h in comments_by_hour:
    avg_by_hour.append([h,comments_by_hour[h]/counts_by_hour[h]])
print(avg_by_hour)    
    

[['10', 13.440677966101696], ['17', 11.46], ['22', 6.746478873239437], ['09', 5.5777777777777775], ['05', 10.08695652173913], ['01', 11.383333333333333], ['13', 14.741176470588234], ['16', 16.796296296296298], ['18', 13.20183486238532], ['00', 8.127272727272727], ['07', 7.852941176470588], ['02', 23.810344827586206], ['20', 21.525], ['19', 10.8], ['12', 9.41095890410959], ['11', 11.051724137931034], ['03', 7.796296296296297], ['06', 9.022727272727273], ['23', 7.985294117647059], ['08', 10.25], ['15', 38.5948275862069], ['04', 7.170212765957447], ['14', 13.233644859813085], ['21', 16.009174311926607]]


## Sort and print values form a List of List

In [17]:
swap_avg_by_hour = []
for i in avg_by_hour:
    swap_avg_by_hour.append([i[1],i[0]])
print(swap_avg_by_hour)
sorted_swap = sorted(swap_avg_by_hour,reverse = True)
print("Top 5 Hours for Ask Post Comments")
print()

for r in sorted_swap[:5]:
    avg = r[0]
    hr = r[1]
    hour = dt.datetime.strptime(hr,"%H").strftime("%H:%M")
    hour_avg_string = "{h}: {a:2f} average comments per post".format(h=hour,a=avg)
    print(hour_avg_string)    
    
    
    

[[13.440677966101696, '10'], [11.46, '17'], [6.746478873239437, '22'], [5.5777777777777775, '09'], [10.08695652173913, '05'], [11.383333333333333, '01'], [14.741176470588234, '13'], [16.796296296296298, '16'], [13.20183486238532, '18'], [8.127272727272727, '00'], [7.852941176470588, '07'], [23.810344827586206, '02'], [21.525, '20'], [10.8, '19'], [9.41095890410959, '12'], [11.051724137931034, '11'], [7.796296296296297, '03'], [9.022727272727273, '06'], [7.985294117647059, '23'], [10.25, '08'], [38.5948275862069, '15'], [7.170212765957447, '04'], [13.233644859813085, '14'], [16.009174311926607, '21']]
Top 5 Hours for Ask Post Comments

15:00: 38.594828 average comments per post
02:00: 23.810345 average comments per post
20:00: 21.525000 average comments per post
16:00: 16.796296 average comments per post
21:00: 16.009174 average comments per post


The hour that received the most comments is the 15:00 hour with an average of about 38.59 comments per post.It is followed by 02:00 which receives about 23.4 comments per post.This implies there is about a 60 percent increase in the rate from highest to second highest value.
15:00 is military time which can be safely converted to 3pm EST time as per data dictionary


## Conclusion

In this project, we analyzed ask posts and show posts to determine which type of post and time receive the most comments on average. Based on our analysis, to maximize the amount of comments a post receives, we'd recommend the post be categorized as ask post and created between 15:00 and 16:00 (3:00 pm est - 4:00 pm est).